### ISSUE TITLE
Unable to build gl_depth_sim package

### ISSUE THREAD

**Issue by lllllyang** [psh117/ljcmp#1]
Hi, I try to follow your steps. I installed all the dependencies, but I face this error during `catkin build`- 

![Screenshot from 2024-07-09 16-01-03](https://github.com/psh117/ljcmp/assets/110092628/0cda3475-7cbe-42a8-8c56-056f1f105375)

Please let us know how to proceed.

**Comment by psh117** [comment]
Hi,

Did you check all files are cloned or downloaded?

`gl_depth_sim:: toCvImage16U(gl_depth_sim::DepthImage const&, cv::Mat&)` function is implemented in `gl_depth_sim/src/interfaces/opencv_interface.cpp`

**Comment by psh117** [comment]
And I think you don't need to build `usage_example.cpp`. Did you modify `CMakeLists.txt`?

**Comment by lllllyang** [comment]
Hi, thanks for your reply.

To clarify, I just explain what I did.
1. I install a clean ubuntu 20.04
2. install [ROS](http://wiki.ros.org/ROS/Installation) (Noetic)
3. install [MoveIt!](https://moveit.ros.org/install/)
4. install pytorch.
5. Follow your steps, then I go to [suhan_robot_model_tools](https://github.com/psh117/suhan_robot_model_tools), follow the readme there.
5.1 
sudo apt install ros-$ROS_DISTRO-combined-robot-hw ros-$ROS_DISTRO-trac-ik ros-noetic-moveit ros-$ROS_DISTRO-nlopt
sudo apt install libnlopt-dev libnlopt-cxx-dev libglfw3-dev
pip install tqdm matplotlib
5.2 
cd ~/catkin_ws/src/
git clone https://github.com/psh117/gl_depth_sim
git clone https://github.com/psh117/suhan_robot_model_tools.git
catkin build
Now the error comes, then I check https://github.com/psh117/gl_depth_sim, I delete the folder "[roscpp_gl_depth_sim_demos](https://github.com/psh117/gl_depth_sim/tree/master/roscpp_gl_depth_sim_demos)" since this is the error and I do not need this demo.
Subsequently, catkin build finally works.
5.3 
cd dir/to/catkin_ws/src/suhan_robot_model_tools
pip install -e .
6. go to [dual_panda_moveit_config](https://github.com/psh117/dual_panda_moveit_config) (for Dual Manipulation), just git clone this, then 
cd dir/to/catkin_ws/src/dual_panda_moveit_config/robot_model
./copy_model.sh
Now another error comes, I check it and I find because I do not have franka robot model, so I just download from ROS. then this aslo works.
7. I git clone https://github.com/psh117/ljcmp.git and then  pip install -e .

However, for the following steps, it is hard to follow and I cannot understand these steps, for example, after I download dataset and model file, where should I put them? Then what is the steps to see the results? Sorry for any inconvenience caused, and thank you for your patience.

**Comment by psh117** [comment]
Sorry for the late response!

Are you still having problems?

After Step 7, you will need to download the dataset and pre-trained models for evaluation.
You may follow benchmark section to see the results.

https://github.com/psh117/ljcmp#benchmark

Download the [benchmarking dataset](https://drive.google.com/file/d/1m6nnp9GOyCLyVnzu--Z5aFgbAtvvQfeu) to dataset directory

Download the [pre-trained models](https://drive.google.com/file/d/1J8qC-ufGLjQHrloTDw12R6vwMlnhFnVD) to model directory

If you have further questions, do not hesitate to reach out to me.

**Comment by lllllyang** [comment]
Hi, thanks for your feedback, my previous problem has been solved. It seems I can run your benchmark and train your model.
![result](https://github.com/user-attachments/assets/28a65c9d-cde4-4999-973d-9abfa17af6e3)

Meanwhile, I have one question

>   when I run 
>   roslaunch dual_panda_moveit_config planning_context.launch load_robot_description:=true
>   and 
>   python script/benchmark.py -E panda_dual
>   in two separate terminals, I cannot see the simulated result (e.g., like the gif in your readme) **Would you mind telling me how to visualize the simulation result**?

Sorry for this simple question, after reading your code and learn rviz a bit, I can visualize your result by running **python3 scripts/benchmark.py -E panda_dual_orientation --display==true**


 1. would you mind showing some diagram to explain how to design these two functions (  **def set_constraint(): and def set_constraint_by_condition(condition)**:) ? 
Sorry but it is hard to understand the constrained equations here...
![constrains](https://github.com/user-attachments/assets/fde86d8e-46ac-4a0c-8c5e-74d0fe1ba934)
Unfortunately, I cannot find related equations in your paper. It could be better if some diagram can explain your constrained equations...

Thanks a lot for your time and patience!

**Comment by psh117** [comment]
What kind of constraints are you trying to set? Are they related to the closed-chain constraint of two manipulators? If you can let me know what the application is, I can provide an example.

The constraint equation in the example is as follows:

$$
(T_0  T_{chain})^{-1} T_1 = I
$$

Technically, the error is calculated separately in position and orientation.

<img width="604" alt="image" src="https://github.com/user-attachments/assets/92bca6a2-90b7-494c-8f27-978a414e798a">

Maybe this figure will help with the calculation of chain_pos and chain_rot. The colors of the end-effector represent the following: X is red, Y is blue, and Z is green.

**Comment by psh117** [comment]
What kind of constraints are you trying to set? Are they related to the closed-chain constraint of two manipulators? If you can let me know what the application is, I can provide an example.

The constraint equation in the closed-chain constraint example is as follows:

(
T
0
T
c
h
a
i
n
)
−
1
T
1
=
I

Technically, the error is calculated separately in position and orientation.

image
Maybe this figure will help with the calculation of chain_pos and chain_rot. The colors of the end-effector represent the following: X is red, Y is blue, and Z is green.

**Comment by psh117** [comment]
What kind of constraints are you trying to set? Are they related to the closed-chain constraint of two manipulators? If you can let me know what the application is, I can provide an example.

The constraint equation in the closed-chain constraint example is as follows:

$$
(T_0 T_{chain})^{-1} T_1 = I
$$

Technically, the error is calculated separately in position and orientation.

<img width="628" alt="image" src="https://github.com/user-attachments/assets/09487680-9eda-4be8-8274-3b663cff7518">

Maybe this figure will help with the calculation of chain_pos and chain_rot. The colors of the end-effector represent the following: X is red, Y is blue, and Z is green.

**Comment by lllllyang** [comment]
Hi, this is exactly the diagram I wanted. Now I know how to set constraints and reuse your code. Your work is really excellent, thank you very much!

**Comment by psh117** [comment]
Thank you for the kind words!

If you have any further questions or issues, please don't hesitate to reach out.
